pithos/plugins/notification_icon.py

```diff
@@ -12,57 +12,119 @@
 # You should have received a copy of the GNU General Public License along
 # with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-import sys
-import gi
+import logging
 
-from gi.repository import GObject, Gdk, Gtk
+from gi.repository import (
+    GObject,
+    GLib,
+    Gtk,
+    Gdk,
+    Gio
+)
 
 from pithos.plugin import PithosPlugin
 
-# Use appindicator if installed
-try:
-    gi.require_version('AppIndicator3', '0.1')
-    from gi.repository import AppIndicator3 as AppIndicator
-    indicator_capable = True
-except (ImportError, ValueError):
-    indicator_capable = False
-
-
-def backend_is_supported(window):
-    if sys.platform in ('win32', 'darwin'):
-        return True
-    display = window.props.screen.get_display()
-    return type(display).__name__.endswith('X11Display')
-
 
 class PithosNotificationIcon(PithosPlugin):
     preference = 'show_icon'
     description = 'Adds pithos icon to system tray'
 
+    appindicator = None
+
     def on_prepare(self):
-        if indicator_capable:
-            self.ind = AppIndicator.Indicator.new("io.github.Pithos-tray",
-                                                  "io.github.Pithos-tray",
-                                                  AppIndicator.IndicatorCategory.APPLICATION_STATUS)
-            # FIXME: AppIndicator might be falling back to XEmbed
-        elif not backend_is_supported(self.window):
-            self.prepare_complete(error='Notification icon requires X11 or AppIndicator')
+        def check_for_x():
+            self.appindicator = None
+            display = self.window.props.screen.get_display()
+            if not type(display).__name__.endswith('X11Display'):
+                logging.debug('Notification icon requires X11 or AppIndicator Support')
+                self.prepare_complete(error='Notification icon requires X11 or AppIndicator Support')
+            else:
+                self.prepare_complete()
+
+        def on_plasmashell_finish(source, result, data):
+            try:
+                plasmashell = Gio.DBusProxy.new_finish(result)
+            except GLib.Error as e:
+                logging.debug(e)
+                check_for_x()
+            else:
+                if plasmashell.get_name_owner():
+                    logging.debug('Notification icon requires AppIndicator on Plasma')
+                    self.prepare_complete(error='Notification icon requires AppIndicator on Plasma')
+                else:
+                    check_for_x()
+
+        def check_for_plasma():
```

Comment:
> I forget if I already asked, but does Plasma not own org.kde.StatusNotifierWatcher ?

appindicator is basically a sh!t wrapper around a DBus service kinda like libnotify except ofc it's the creator not the consumer. KDE/Qt tray icons work in Unity with little to no modification without the aid of appindicator if I remember correctly.